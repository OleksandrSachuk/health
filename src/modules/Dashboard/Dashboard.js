import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  airPressure: PropTypes.number.isRequired,
  humidity: PropTypes.number.isRequired,
  temperature: PropTypes.number.isRequired,
};

const defaultProps = {};

const Dashboard = ({
                     airPressure,
                     humidity,
                     temperature,
                   }) =>
  (
    <div>
      <div>
        Temperature: { temperature } Â°C
      </div>
      <div>
        Air pressure: { airPressure } mmHg
      </div>
      <div>
        Humidity: { humidity } %
      </div>
    </div>
  );

Dashboard.propTypes = propTypes;
Dashboard.defaultProps = defaultProps;

export default Dashboard;
