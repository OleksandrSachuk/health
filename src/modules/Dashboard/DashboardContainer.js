import { of, combineLatest } from 'rxjs';
import { delay, repeat, concatMap, timeInterval } from 'rxjs/operators';
import { compose, lifecycle, withState } from 'recompose';
import random from 'lodash/random';

import Dashboard from './Dashboard';

export default compose(
  withState('temperature', 'setTemperature', 0),
  withState('airPressure', 'setAirPressure', 0),
  withState('humidity', 'setHumidity', 0),

  lifecycle({

    componentDidMount() {
      const randomData = ({ min, max, float, decimal }) => Math.round(random(min, max, float) * decimal) / decimal;

      const randomTemperatureData = {
        min: 36,
        max: 37,
        float: true,
        decimal: 10,
      };

      const randomAirPresureData = {
        min: 740,
        max: 750,
        float: false,
        decimal: 1000,
      };

      const randomHumidityData = {
        min: 40,
        max: 60,
        float: false,
        decimal: 100,
      };

      const {
        setTemperature,
        setAirPressure,
        setHumidity,
      } = this.props;

      const randomDelay = (bottom, top) => Math.floor(Math.random() * (1 + top - bottom)) + bottom;

      this.temperature$ = of(randomData(randomTemperatureData))
        .pipe(
          concatMap(() => of(randomData(randomTemperatureData))
            .pipe(
              delay(randomDelay(100, 1500)),
            ),
          ),
          repeat(),
          timeInterval(),
        );

      this.airPresure$ = of(randomData(randomAirPresureData))
        .pipe(
          concatMap(() => of(randomData(randomAirPresureData))
            .pipe(
              delay(randomDelay(100, 200)),
            ),
          ),
          repeat(),
          timeInterval(),
        );

      this.humidity$ = of(randomData(randomHumidityData))
        .pipe(
          concatMap(() => of(randomData(randomHumidityData))
            .pipe(
              delay(randomDelay(100, 200)),
            ),
          ),
          repeat(),
          timeInterval(),
        );

      this.summary$ = combineLatest(this.temperature$, this.airPresure$, this.humidity$)
        .subscribe(([temperature, airPresure, humidity]) => {

          const missedValue = 'N/A';

          const { value: temperatureValue, interval: temperatureInterval } = temperature;
          temperatureInterval < 1000 ? setTemperature(temperatureValue) : setTemperature(missedValue);

          const { value: airPresureValue, interval: airPresureInterval } = airPresure;
          airPresureInterval < 1000 ? setAirPressure(airPresureValue) : setAirPressure(missedValue);

          const { value: humidityValue, interval: humidityInterval } = humidity;
          humidityInterval < 1000 ? setHumidity(humidityValue) : setHumidity(missedValue);

        });
    },

    componentWillUnmount() {
      this.summary$.unsubscribe();
    },

  }),
)(Dashboard);

// , (a,b,c)=> console.log('a,b,c', a,b,c)