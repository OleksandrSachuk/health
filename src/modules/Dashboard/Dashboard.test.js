import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Dashboard from './';

it('renders Dashboard without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Dashboard />, div);
  ReactDOM.unmountComponentAtNode(div);
});


it('Dashboard snapshot test', () => {

  const component = renderer.create(
    <Dashboard />,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
