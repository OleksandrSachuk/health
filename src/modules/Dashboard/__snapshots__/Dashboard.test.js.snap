// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard snapshot test 1`] = `
<table
  className="MuiTable-root-21"
>
  <thead
    className="MuiTableHead-root-22"
  >
    <tr
      className="MuiTableRow-root-23 MuiTableRow-head-26"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-head-29"
        scope="col"
      >
        Name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-head-29 Dashboard-tableCellCenter-20"
        scope="col"
      >
        Value
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-head-29 Dashboard-tableCellCenter-20"
        scope="col"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-36"
  >
    <tr
      className="MuiTableRow-root-23"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-body-30"
        scope="row"
      >
        Temperature
      </th>
      <td
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-body-30 MuiTableCell-numeric-32 Dashboard-tableCellCenter-20"
        scope={undefined}
      >
        0
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-body-30 Dashboard-tableCellCenter-20"
        scope={undefined}
      >
        Â°C
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-23"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-body-30"
        scope="row"
      >
        Air pressure
      </th>
      <td
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-body-30 MuiTableCell-numeric-32 Dashboard-tableCellCenter-20"
        scope={undefined}
      >
        0
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-body-30 Dashboard-tableCellCenter-20"
        scope={undefined}
      >
        mmHg
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-23"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-body-30"
        scope="row"
      >
        Humidity
      </th>
      <td
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-body-30 MuiTableCell-numeric-32 Dashboard-tableCellCenter-20"
        scope={undefined}
      >
        0
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-28 MuiTableCell-body-30 Dashboard-tableCellCenter-20"
        scope={undefined}
      >
        %
      </td>
    </tr>
  </tbody>
</table>
`;
